mentat/include_files.py

```diff
@@ -44,104 +23,224 @@ def is_file_text_encoded(abs_path: Path):
         return False
 
 
-def abs_files_from_list(paths: list[Path], check_for_text: bool = True):
-    """Returns a set of CodeFiles from a list of paths."""
-    files_direct = set[CodeFeature]()
-    file_paths_from_dirs = set[Path]()
-    invalid_paths = list[str]()
-    for path in paths:
-        file = CodeFeature(os.path.realpath(path))
-        path = Path(file.path)
-        if path.is_file():
-            if check_for_text and not is_file_text_encoded(path):
-                logging.info(f"File path {path} is not text encoded.")
-                invalid_paths.append(str(path))
-            else:
-                files_direct.add(file)
-        elif path.is_dir():
-            nonignored_files = set(
-                map(
-                    lambda f: Path(os.path.realpath(path / f)),
-                    get_non_gitignored_files(path),
-                )
-            )
+def validate_and_format_path(
+    path: Path | str, cwd: Path, check_for_text: bool = True
+) -> Path:
+    """Validate and format a path.
+
+    Args:
+        `path` - A file path, file interval path, directory path, or a glob pattern
+        `check_for_text` - Check if the file can be opened. Default to True
 
-            file_paths_from_dirs.update(
-                filter(
-                    lambda f: (not check_for_text) or is_file_text_encoded(f),
-                    nonignored_files,
-                )
+    Return:
+        An absolute path
+    """
+    path = Path(path)
+
+    # Get absolute path
+    if path.is_absolute():
+        abs_path = path
+    else:
+        abs_path = cwd.joinpath(path)
+
+    # Resolve path (remove any '..')
+    abs_path = abs_path.resolve()
+
+    # Validate different path types
+    # File
+    if abs_path.is_file():
+        if not abs_path.exists():
+            raise PathValidationError(f"File {abs_path} does not exist")
+        if check_for_text and not is_file_text_encoded(abs_path):
+            raise PathValidationError(f"Unable to read file {abs_path}")
+    # File interval
+    elif len(str(abs_path).rsplit(":", 1)) > 1:
+        _interval_path, interval_str = str(abs_path).rsplit(":", 1)
+        interval_path = Path(_interval_path)
+        if not interval_path.exists():
+            raise PathValidationError(f"File {interval_path} does not exist")
+        if check_for_text and not is_file_text_encoded(interval_path):
+            raise PathValidationError(f"Unable to read file {interval_path}")
+        intervals = parse_intervals(interval_str)
+        if len(intervals) == 0:
+            raise PathValidationError(
+                f"Unable to parse intervals for path {interval_path}"
             )
+    # Directory
+    elif abs_path.is_dir():
+        pass
+    # Glob pattern
+    elif re.search(r"[\*\?\[\]]", str(path)):
```

Comment:
Why not just check if glob.glob returns anything?

Reply:
Because calling `glob.glob` might return a huge amount of files unnecessarily. Here we are just doing a no-op check of `path` containing valid glob pattern characters

Reply:
Actually just remembered `glob.iglob` is a thing. Just added that for more robust validation